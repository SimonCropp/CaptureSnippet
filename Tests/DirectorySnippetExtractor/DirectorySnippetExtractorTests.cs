using System;
using System.Collections.Generic;
using System.IO;
using ApprovalTests.Reporters;
using CaptureSnippets;
using NuGet.Versioning;
using NUnit.Framework;
using ObjectApproval;

[TestFixture]
[UseReporter(typeof(AllFailingTestsClipboardReporter), typeof(DiffReporter))]
public class DirectorySnippetExtractorTests
{
    [Test]
    public void VerifyLambdasAreCalled()
    {
        var targetDirectory = Path.Combine(TestContext.CurrentContext.TestDirectory, "DirectorySnippetExtractor");
        var snippetMetaData = new SnippetMetaData(VersionRange.All, "package");
        var result = new TestResult();
        var extractor = new DirectorySnippetExtractor(
            extractMetaDataFromPath: (rootDirectory, path, parent) =>
            {
                result.ExtractMetaDatas.Add(new CapturedExtractMetaData {RootDirectory = rootDirectory, Path = path, Parent = parent});
                return snippetMetaData;
            },
            includeDirectory: path =>
            {
                result.IncludeDirectories.Add(new CapturedIncludeDirectory {Path = path});
                return true;
            },
            includeFile: path =>
            {
                result.IncludeFiles.Add(new CapturedIncludeFile {Path = path});
                return true;
            },
            translatePackage: (path, alias) =>
            {
                result.TranslatePackages.Add(new CapturedTranslatePackage {Path = path, Alias = alias});
                return alias;
            },
            parseVersion: (version, path, metaDataForParentPath) =>
            {
                result.ParseVersions.Add(new CapturedParseVersion {Version = version, Path = path, MetaDataForParentPath = metaDataForParentPath});
                return VersionRange.All;
            });
        extractor.FromDirectory(targetDirectory)
            .GetAwaiter()
            .GetResult();
        ObjectApprover.VerifyWithJson(result, s => s.Replace(@"\\",@"\").Replace(targetDirectory, @"root\"));
    }

    public class TestResult
    {
        public Sorted<CapturedExtractMetaData> ExtractMetaDatas = new Sorted<CapturedExtractMetaData>(_ => _.Path);
        public Sorted<CapturedIncludeDirectory> IncludeDirectories = new Sorted<CapturedIncludeDirectory>(_ => _.Path);
        public Sorted<CapturedIncludeFile> IncludeFiles = new Sorted<CapturedIncludeFile>(_ => _.Path);
        public Sorted<CapturedTranslatePackage> TranslatePackages = new Sorted<CapturedTranslatePackage>(_ => _.Path);
        public Sorted<CapturedParseVersion> ParseVersions = new Sorted<CapturedParseVersion>(_ => _.Path);
    }

    public class Sorted<T> : SortedSet<T>
    {
        public Sorted(Func<T,string> func):base(Comparer<T>.Create((data1, data2) => func(data1).CompareTo(func(data2))))
        {

        }
    }

    public class CapturedExtractMetaData
    {
        public SnippetMetaData Parent;
        public string Path;
        public string RootDirectory;
    }

    public class CapturedIncludeDirectory
    {
        public string Path;
    }

    public class CapturedIncludeFile
    {
        public string Path;
    }

    public class CapturedTranslatePackage
    {
        public string Path;
        public string Alias;
    }

    public class CapturedParseVersion
    {
        public string Version;
        public string Path;
        public SnippetMetaData MetaDataForParentPath;
    }
}