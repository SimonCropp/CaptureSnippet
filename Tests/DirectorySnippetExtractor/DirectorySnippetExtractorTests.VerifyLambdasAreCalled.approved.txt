{
  "ExtractMetaDatas": [
    {
      "Parent": null,
      "Path": "root\",
      "RootDirectory": "root\"
    },
    {
      "Parent": {
        "Version": "all",
        "Package": "package"
      },
      "Path": "root\\1.1",
      "RootDirectory": "root\"
    },
    {
      "Parent": {
        "Version": "all",
        "Package": "package"
      },
      "Path": "root\\1.1\2.2",
      "RootDirectory": "root\"
    },
    {
      "Parent": {
        "Version": "all",
        "Package": "package"
      },
      "Path": "root\\1.1\2.2\code",
      "RootDirectory": "root\"
    },
    {
      "Parent": {
        "Version": "all",
        "Package": "package"
      },
      "Path": "root\\1.1\code",
      "RootDirectory": "root\"
    }
  ],
  "IncludeDirectories": [
    {
      "Path": "root\"
    },
    {
      "Path": "root\\1.1"
    },
    {
      "Path": "root\\1.1\2.2"
    }
  ],
  "IncludeFiles": [
    {
      "Path": "root\\1.1\2.2\code.txt"
    },
    {
      "Path": "root\\1.1\code.txt"
    }
  ],
  "TranslatePackages": [
    {
      "Path": "root\\1.1\2.2\code.txt",
      "Alias": "package2"
    },
    {
      "Path": "root\\1.1\code.txt",
      "Alias": "package1"
    }
  ],
  "ParseVersions": [
    {
      "Version": "2.0",
      "Path": "root\\1.1\2.2\code.txt",
      "MetaDataForParentPath": {
        "Version": "all",
        "Package": "package"
      }
    },
    {
      "Version": "1.0",
      "Path": "root\\1.1\code.txt",
      "MetaDataForParentPath": {
        "Version": "all",
        "Package": "package"
      }
    }
  ]
}